sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java

```diff
@@ -585,11 +585,23 @@ protected void initialize(SchedulerDriver driver) throws InterruptedException {
     }
 
     private void killUnneededTasks(Set<String> taskToDeployNames) {
-        Set<Protos.TaskID> taskIds = stateStore.fetchTasks().stream()
+        Set<Protos.TaskInfo> taskInfos = stateStore.fetchTasks().stream()
                 .filter(taskInfo -> !taskToDeployNames.contains(taskInfo.getName()))
+                .collect(Collectors.toSet());
+
+        Set<Protos.TaskID> taskIds = taskInfos.stream()
                 .map(taskInfo -> taskInfo.getTaskId())
                 .collect(Collectors.toSet());
 
+        // Clear the TaskIDs from the TaskInfos so we drop all future TaskStatus Messages
+        Set<Protos.TaskInfo> cleanedTaskInfos = taskInfos.stream()
+                .map(taskInfo -> taskInfo.toBuilder())
+                .map(builder -> builder.setTaskId(Protos.TaskID.newBuilder().setValue("")).build())
+                .collect(Collectors.toSet());
+
+        cleanedTaskInfos.forEach(taskInfo -> stateStore.clearTask(taskInfo.getName()));
```

Comment:
I think if we failed to kill it (e.g. scheduler restarted just before) then the running status would be restored in task reconciliation. So it wouldn't go completely missing but it may end up in the task still running (unless there's something that would garbage collect it at that point..)